hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go

```diff
@@ -593,11 +594,24 @@ func (r *HostedClusterReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 		return ctrl.Result{}, nil
 	}
 
-	// Default the infraID if unset
-	if hcluster.Spec.InfraID == "" {
-		hcluster.Spec.InfraID = infraid.New(hcluster.Name)
-		if err := r.Update(ctx, hcluster); err != nil {
-			return ctrl.Result{}, fmt.Errorf("failed to update hostedcluster after defaulting the InfraID: %w", err)
+	{
+		// Default the ClusterID if unset
+		needsUpdate := false
+		if hcluster.Spec.ClusterID == "" {
+			hcluster.Spec.ClusterID = uuid.NewString()
+			needsUpdate = true
```

Comment:
can all this logic be unit tested?

Also if the uuid is not generated it should be validated with e.g uuid.Parse() and signal failure in a condition see `validateConfigAndClusterCapabilities`.

Ideally additionally we would also fail before persisting via openapi regexp validation https://book.kubebuilder.io/reference/markers/crd-validation.html / https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules